perspective_reddit_bot/moderate_subreddit.py

```diff
@@ -0,0 +1,152 @@
+"""A reddit bot which uses the Perpsective API to help moderate subreddits.
+
+Args:
+      subreddit: (required) A positional argument for the name of the subreddit to moderate.
+      -rule_config_file: (optional) The file which contains moderation rules. Defaults to 'rules.yaml'
+      -output_dir: (optional) If set, scores for all streamed commments are written to a file in this directory.
+
+Copyright 2018 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import argparse
+import os
+import time
+
+from creds import creds
+import pandas as pd
+import perspective_client
+from perspective_rule import Rule
+import praw
+import score_dataset
+import yaml
+
+TEXT_COLUMN = 'comment_text'
+
+LANGUAGE = 'en'
+# TODO(nthain): support automated language detection.
+
+
+def parse_rules(filepath):
+  rules = []
+  models = []
+  with open(filepath) as f:
+    raw_rules = yaml.load_all(f)
+    for r in raw_rules:
+      rules.append(Rule(r['perspective_score'],
+                        r['action'],
+                        r.get('report_reason')))
+      models.extend(r['perspective_score'].keys())
+  assert len(rules) > 0, 'Rules file empty!'
+  return rules, models
+
+
+def bot_is_mod(reddit, subreddit):
+  mods = subreddit.moderator()
+  return reddit.user.me() in mods
+
+
+def score_subreddit(creds_dict,
+                    subreddit_name,
+                    rules,
+                    models,
+                    output_path=None):
+  """Scores a stream of commments via the perspective API and applies pre-defined moderation rules.
+
+  Args:
+    creds_dict: (dict) A dictionary of API credentials for Perspective and Reddit.
+    subreddit_name: (str) The name of the subreddit to stream.
+    rules: (list) A list of rules to apply to each comment.
+    models: (list) A list of models that the API must call to apply rules.
+    output_path: (str, optional) If supplied, all comments and scores will be written to this path.
+  """
+
+  reddit = praw.Reddit(client_id=creds_dict['reddit_client_id'],
+                       client_secret=creds_dict['reddit_client_secret'],
+                       user_agent=creds_dict['reddit_user_agent'],
+                       username=creds_dict['reddit_username'],
+                       password=creds_dict['reddit_password'])
+  subreddit = reddit.subreddit(subreddit_name)
+
+  try:
+    mod_permissions = bot_is_mod(reddit, subreddit)
+  except Exception:
+    mod_permissions = False
```

Comment:
When does this raise exceptions? I think it'd be better to push the exception handling into the `bot_is_mod` helper, so it's guaranteed to just return a bool.

Reply:
Done.